README.md
## grblHAL ##

__NEWS!__ grblHAL is moving to a [new home](https://github.com/grblHAL) where the project is split into separate repositories, one for the shared core, one for each plugin and one for for each driver. The driver dependencies to the core and plugins are set up as git submodules and drivers can thus be downloaded separately.

There are both upsides and downsides to this move, the most notable downside for users is that downloading the source code as a .zip-file no-longer get all dependencies resolved.
Each submodule a driver is dependent on has to be downloaded separately and copied to the corresponding folder in the driver
This is a bit more involved than the copying that was needed in previous releases where the dependent code can be found in the downloaded .zip-file.

An upside is that a driver download with [git](https://git-scm.com/downloads) or [GitHub desktop](https://desktop.github.com/) will fetch all dependencies automatically and the download is no longer "polluted" with unwanted drivers.

For the time beeing I will keep this repository intact and the source code in sync.
New issues should be raised, pull requests submitted and discussions started at the new home.
Use the [core repository](https://github.com/grblHAL/core) for raising new issues or starting discussions if it is not clear to you where they belong. 

This repository will eventually be moved to the new home and kept as an reference.

---

grblHAL has [many extensions](https://github.com/terjeio/grblHAL/wiki) that may cause issues with some senders. As a workaround for these a [compile time option](https://github.com/terjeio/grblHAL/wiki/Changes-from-grbl-1.1#workaround) has been added that disables extensions selectively. 

__IMPORTANT!__ grblHAL defaults to normally closed \(NC\) switches for inputs, if none are connected when testing it is likely that the controller will start in alarm mode.  
Temporarily short the Reset, E-Stop and Safety Door<sup>4</sup> inputs to ground or invert the corresponding inputs by setting `$14=73` to avoid that.  
Please check out [this Wiki page](https://github.com/terjeio/grblHAL/wiki/Changes-from-grbl-1.1) for additional important information.

Windows users may try [ioSender](https://github.com/terjeio/Grbl-GCode-Sender), binary releases can be found [here](https://github.com/terjeio/Grbl-GCode-Sender/releases).
It has been written to complement grblHAL and has features such as proper keyboard jogging, advanced probing, automatic reconfiguration of DRO display for up to 6 axes, lathe mode including conversational G-Code generation, 3D rendering, macro support etc. etc.

---

Latest build date is 20210223, see the [changelog](https://github.com/terjeio/grblHAL/blob/master/changelog.md) for details.

---

grblHAL is a no-compromise, high performance, low cost alternative to parallel-port-based motion control for CNC milling and is based on the [Arduino version of grbl](https://github.com/gnea/grbl). It is mainly aimed at ARM processors \(or other 32-bit MCUs\) with ample amounts of RAM and flash \(compared to AVR 328p\) and requires a [hardware driver](drivers/ReadMe.md) to be functional.
Currently drivers are available for 13 different processors/processor families all of which share the same core.

grblHAL has an open architecture allowing [plugins](plugins/README.md) to extend functionality.
User made plugins can be added to grblHAL without changing a single file in the source<sup>1</sup>, and allows for a wide range extensions to be added.
New M-codes can be added, space for plugin specific settings can be allocated, events can be subscribed to etc. etc.  
Adding code to drive an ATC, extra outputs or even adding a UI<sup>2</sup> has never been easier. You can even add your own [driver](templates/arm-driver) if you feel so inclined.

HAL = Hardware Abstraction Layer

The controller is written in highly optimized C utilizing features of the supported processors to achieve precise timing and asynchronous operation.
It is able to maintain up to 300kHz<sup>3</sup> of stable, jitter free control pulses.

It accepts standards-compliant g-code and has been tested with the output of several CAM tools with no problems. Arcs, circles and helical motion are fully supported, as well as, all other primary g-code commands. Macro functions, variables, and some canned cycles are not supported, but we think GUIs can do a much better job at translating them into straight g-code anyhow.

Grbl includes full acceleration management with look ahead. That means the controller will look up to 16 motions into the future and plan its velocities ahead to deliver smooth acceleration and jerk-free cornering.

This is a port/rewrite of [grbl 1.1f](https://github.com/gnea/grbl) and should be compatible with GCode senders compliant with the specifications for that version. It should be possible to change default compile-time configurations if problems arise, eg. the default serial buffer sizes has been increased in some of the [drivers](drivers/ReadMe.md) provided.

Check out [grbl.org](https://www.grbl.org/) for build showcases, blog posts etc.

<sup>1</sup> This feature is only to be used for private plugins, if shared then a single call must be added to the driver code of the target processors.   
<sup>2</sup> I do not usually recommend doing this, and I will not accept pull requests for any. However I may add a link to the github repository for any that might be made.  
<sup>3</sup> Driver/processor dependent.  
<sup>4</sup> Not enabled by default if building from source, but may be enabled in prebuilt firmware.

***

```
List of Supported G-Codes:
  - Non-Modal Commands: G4, G10L2, G10L20, G28, G30, G28.1, G30.1, G53, G92, G92.1
  - Additional Non-Modal Commands: G10L1*, G10L10*, G10L11*
  - Motion Modes: G0, G1, G2, G3, G5, G38.2, G38.3, G38.4, G38.5, G80, G33*
  - Canned cycles: G73, G81, G82, G83, G85, G86, G89, G98, G99
  - Repetitive cycles: G76*
  - Feed Rate Modes: G93, G94, G95*, G96*, G97*
  - Unit Modes: G20, G21
  - Scaling: G50, G51
  - Lathe modes: G7*, G8*
  - Distance Modes: G90, G91
  - Arc IJK Distance Modes: G91.1
  - Plane Select Modes: G17, G18, G19
  - Tool Length Offset Modes: G43*, G43.1, G43.2*, G49
  - Cutter Compensation Modes: G40
  - Coordinate System Modes: G54, G55, G56, G57, G58, G59, G59.1, G59.2, G59.3
  - Control Modes: G61
  - Program Flow: M0, M1, M2, M30, M60
  - Coolant Control: M7, M8, M9
  - Spindle Control: M3, M4, M5
  - Tool Change: M6* (Two modes possible: manual** - supports jogging, ATC), M61
  - Switches: M49, M50, M51, M53
  - Output control***: M62, M63, M64, M65, M66, M67, M68
  - Valid Non-Command Words: A*, B*, C*, F, H*, I, J, K, L, N, P, Q*, R, S, T, X, Y, Z

  *  driver/configuration dependent.
  ** requires compatible GCode sender due to protocol extensions, new state and RT command.
  *** number of outputs supported dependent on driver implementation.
```

---
2021-02-25


changelog.md
## grblHAL changelog

Build 20210515:

* Fixed G43 "bug" by changing to LinuxCNC specification. NIST specification is ambiguous.
* Added acceleration override handling, to be used by OpenPNP plugin M-code.

Build 20210505:

* Some bug fixes: relative canned cycles repeat error, comment handling...
* Check mode changes to allow use with SD card streaming.
* Internal buffer changes for reducing RAM footprint. Some HAL extensions.

Build 20210314:

* Added support for tool change protocol to networking protocols using shared stream buffer.
* Added `$S` system command for toggling single step mode, when enabled cycle start has to be issued after each block.
* Some bug fixes.

Build 20210207:

* Added `#define BOARD_MY_MACHINE` option in _my_machine.h_ for building using _my_machine-map.c_, this for simpler handling of user defined pin mappings.  
_my_machine-map.c_ is __*not*__ part of the distributed source and must by added to the project by the user before enabled, typically by copying an existing map file.
* Added HAL layer on top of Trinamic driver low-level code. Unified Triniamic plugins into single plugin.  
This is still work in progress, testers wanted.
* Added core support for up to four limit switches per axis.  
Added `$LEV` command for outputting report containing which limit or control switch(es) caused the last event.  
Report format:  
`[LASTEVENTS:<control signals>,<min>,<max>,<min2>,<max2)]`  
Where `<control signals>` field may contain controls signal letters (`H`, `S` etc.) and the rest axis letters for the corresponding limit switches inputs.
* More settings susbsystem changes and refactored $-system commands parser.  
There are some API changes related to this that may affect user defined plugins.
* Added $7 setting for option "Spindle off with zero speed".
* Added `|FW:grblHAL` element to full real-time report requested by sending `0x87`. 
* Simplified settings handling.
* No longer sends any messages to networking streams on connect.
* Added [driver](drivers/STM32F3xx) for STM32F303 based Blackpills.
* Updated support for Trinamic TMC5160 drivers, currently for the SKR 1.x boards \(LPC176x driver\) - testing i progress.
* Initial support for Trinamic TMC5160 drivers added, currently for the SKR 1.x boards \(LPC176x driver\) - not yet tested.
* Added handling for motor fault signal to the core, similar to E-stop handling. Added new alarm and error code for this.
* Changed to clear homed status on a soft reset only if machine was in motion. Added setting flag for always keeping homed status on soft reset to `$22` setting.  
__NOTE:__ This change is experimental and might be changed or reverted. Please report any problems related to this.
* Updated gcode parser \(grbl/gcode.c\) to use bitfields structs instead of bitfield variables.  
Done to improve readability and for easier debugging.  
Note that this is a major change and there is a non-zero risk that mistakes has been made.
* Changed signature of user mcode validation function to use a bitfield union for value words available.  
Removed the need for user mcode parameter words to have an associated value. This means that [user mcode](https://github.com/terjeio/grblHAL/tree/test/templates) implementations now must check this locally.  
If no associated value is provided the corresponding value in the value struct is set to `NAN` (Not A Number) for floats and all bits set to 1 for integers.
* Refactored Trinamic driver code, added initial support for TMC2209. Work in progress.
* Removed the need to copy the core grbl and plugin code to the driver chosen, this is now kept in sync with the master Subversion repository automatically.
* Added initial support for RADDS 1.6 board to SAM3X8E driver \(Arduino Due\). Untested!
* Added C-axis support to iMXRT1062 driver \(Teensy 4.x\).  
Untested and none of the current board maps has the needed pins defined.
* Added alarm and error message for power on self-test \(POS\) failure.  
If POS fails only $-commands are accepted.
* Work in good progress for Trinamic TMC2209 driver support \(UART mode\).  
Processor/board specific driver code has to be added for this, currently testing with STM32F446 and Nucleo-64 breakout board.
* Renumbered setting groups for more logical sorting (by id).
* Harmonized probing code across drivers for planned future extensions.
* Added additional I/O support for the [Teensy 4.1 T41U5XBB board](https://github.com/phil-barrett/grbl-teensy-4) \(iMXRT1062 driver\), 3 outputs and 4 inputs available via `M62` - `M66`.  
Not that the result from reading inputs with `M66` cannot be used in a gcode program in any meaningful way.
* Fixed excessive step pulse jitter in STM32F4xx driver.
* Added [more options](https://github.com/terjeio/grblHAL/wiki/Report-extensions#controller-information-extensions) to the `NEWOPT` tag in the extended `$I` report.
* Error 7 is no longer issued on startup if non-volatile storage \(Flash/EEPROM/FRAM\) for settings is not available.
* [Alarm substate](https://github.com/terjeio/grblHAL/wiki/Report-extensions#realtime-report) \(if available\) is always added to the real-time report if a [complete report](https://github.com/terjeio/grblHAL/wiki/For-sender-developers#single-character-real-time-commands) is requested by sending `0x87`.
* Added input signal and handling for limit switches override.  
The pin is pulled high and requires a normally open \(NO\) push switch for operation. When closed limit pins are excluded from the status report and alarm 12 will not be raised if a limit switch is asserted _on a soft reset_ when "Hard limits" and "Strict mode" is enabled with `$21`.
This allows normal operation so that a manual pull-off can be done before e.g. homing the machine.  

GRBL/gcode.h
/*
  gcode.h - rs274/ngc parser.
  Part of Grbl

  Copyright (c) 2017-2019 Terje Io
  Copyright (c) 2011-2016 Sungeun K. Jeon for Gnea Research LLC
  Copyright (c) 2009-2011 Simen Svale Skogsrud

  Grbl is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  Grbl is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with Grbl.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef gcode_h
#define gcode_h

#include "coolant_control.h"
#include "spindle_control.h"


// Define Grbl status codes. Valid values (0-255)
typedef enum {
    Status_OK = 0,
    Status_ExpectedCommandLetter = 1,
    Status_BadNumberFormat = 2,
    Status_InvalidStatement = 3,
    Status_NegativeValue = 4,
    Status_SettingDisabled = 5,
    Status_SettingStepPulseMin = 6,
    Status_SettingReadFail = 7,
    Status_IdleError = 8,
    Status_SystemGClock = 9,
    Status_SoftLimitError = 10,
    Status_Overflow = 11,
    Status_MaxStepRateExceeded = 12,
    Status_CheckDoor = 13,
    Status_LineLengthExceeded = 14,
    Status_TravelExceeded = 15,
    Status_InvalidJogCommand = 16,
    Status_SettingDisabledLaser = 17,
    Status_Reset = 18,
    Status_NonPositiveValue = 19,

    Status_GcodeUnsupportedCommand = 20,
    Status_GcodeModalGroupViolation = 21,
    Status_GcodeUndefinedFeedRate = 22,
    Status_GcodeCommandValueNotInteger = 23,
    Status_GcodeAxisCommandConflict = 24,
    Status_GcodeWordRepeated = 25,
    Status_GcodeNoAxisWords = 26,
    Status_GcodeInvalidLineNumber = 27,
    Status_GcodeValueWordMissing = 28,
    Status_GcodeUnsupportedCoordSys = 29,
    Status_GcodeG53InvalidMotionMode = 30,
    Status_GcodeAxisWordsExist = 31,
    Status_GcodeNoAxisWordsInPlane = 32,
    Status_GcodeInvalidTarget = 33,
    Status_GcodeArcRadiusError = 34,
    Status_GcodeNoOffsetsInPlane = 35,
    Status_GcodeUnusedWords = 36,
    Status_GcodeG43DynamicAxisError = 37,
    Status_GcodeIllegalToolTableEntry = 38,
    Status_GcodeValueOutOfRange = 39,
    Status_GcodeToolChangePending = 40,
    Status_GcodeSpindleNotRunning = 41,
    Status_GcodeIllegalPlane = 42,
    Status_GcodeMaxFeedRateExceeded = 43,
    Status_GcodeRPMOutOfRange = 44,
    Status_LimitsEngaged = 45,
    Status_HomingRequired = 46,
    Status_GCodeToolError = 47,

    Status_EStop = 50,
    Status_Unhandled = 59, // For internal use only

// Some error codes as defined in bdring's ESP32 port
    Status_SDMountError = 60,
    Status_SDReadError = 61,
    Status_SDFailedOpenDir = 62,
    Status_SDDirNotFound = 63,
    Status_SDFileEmpty = 64,

    Status_BTInitError = 70
} status_code_t;


// Define modal group internal numbers for checking multiple command violations and tracking the
// type of command that is called in the block. A modal group is a group of g-code commands that are
// mutually exclusive, or cannot exist on the same line, because they each toggle a state or execute
// a unique motion. These are defined in the NIST RS274-NGC v3 g-code standard, available online,
// and are similar/identical to other g-code interpreters by manufacturers (Haas,Fanuc,Mazak,etc).
// NOTE: Modal group define values must be sequential and starting from zero.
typedef enum {
    ModalGroup_G0 = 0,  // [G4,G10,G28,G28.1,G30,G30.1,G53,G92,G92.1] Non-modal
    ModalGroup_G1,      // [G0,G1,G2,G3,G33,G38.2,G38.3,G38.4,G38.5,G76,G80] Motion
    ModalGroup_G2,      // [G17,G18,G19] Plane selection
    ModalGroup_G3,      // [G90,G91] Distance mode
    ModalGroup_G4,      // [G91.1] Arc IJK distance mode
    ModalGroup_G5,      // [G93,G94,G95] Feed rate mode
    ModalGroup_G6,      // [G20,G21] Units
    ModalGroup_G7,      // [G40] Cutter radius compensation mode. G41/42 NOT SUPPORTED.
    ModalGroup_G8,      // [G43,G43.1,G49] Tool length offset
    ModalGroup_G10,     // [G98,G99] Return mode in canned cycles
    ModalGroup_G11,     // [G50,G51] Scaling
    ModalGroup_G12,     // [G54,G55,G56,G57,G58,G59] Coordinate system selection
    ModalGroup_G13,     // [G61] Control mode
    ModalGroup_G14,     // [G96,G97] Spindle Speed Mode
    ModalGroup_G15,     // [G7,G8] Lathe Diameter Mode

    ModalGroup_M4,      // [M0,M1,M2,M30] Stopping
    ModalGroup_M6,      // [M6] Tool change
    ModalGroup_M7,      // [M3,M4,M5] Spindle turning
    ModalGroup_M8,      // [M7,M8,M9] Coolant control
    ModalGroup_M9,      // [M49,M50,M51,M53,M56] Override control
    ModalGroup_M10,     // User defined M commands
} modal_group_t;

// Define parameter word mapping.
typedef enum {
    Word_E = 0,
	Word_F,
    Word_H,
    Word_I,
    Word_J,
    Word_K,
    Word_L,
    Word_N,
    Word_P,
    Word_R,
    Word_S,
    Word_T,
    Word_X,
    Word_Y,
    Word_Z,
    Word_Q,
#if N_AXIS > 3
	Word_A,
	Word_B,
	Word_C,
#endif
    Word_D
} parameter_word_t;

#if N_AXIS == 3
#define AXIS_WORDS_MASK ((1 << Word_X)|(1 << Word_Y)|(1 << Word_Z))
#elif N_AXIS == 4
#define AXIS_WORDS_MASK ((1 << Word_X)|(1 << Word_Y)|(1 << Word_Z)|(1 << Word_A))
#elif N_AXIS == 5
#define AXIS_WORDS_MASK ((1 << Word_X)|(1 << Word_Y)|(1 << Word_Z)|(1 << Word_A)|(1 << Word_B))
#else
#define AXIS_WORDS_MASK ((1 << Word_X)|(1 << Word_Y)|(1 << Word_Z)|(1 << Word_A)|(1 << Word_B)|(1 << Word_C))
#endif

typedef union {
    parameter_word_t parameter;
    modal_group_t group;
} word_bit_t;

// Define command actions for within execution-type modal groups (motion, stopping, non-modal). Used
// internally by the parser to know which command to execute.
// NOTE: Some values are assigned specific values to make g-code state reporting and parsing
// compile a litte smaller. Although not
// ideal, just be careful with values that state 'do not alter' and check both report.c and gcode.c
// to see how they are used, if you need to alter them.

// Modal Group G0: Non-modal actions
typedef enum {
    NonModal_NoAction = 0,                  // (Default: Must be zero)
    NonModal_Dwell = 4,                     // G4 (Do not alter value)
    NonModal_SetCoordinateData = 10,        // G10 (Do not alter value)
    NonModal_GoHome_0 = 28,                 // G28 (Do not alter value)
    NonModal_SetHome_0 = 38,                // G28.1 (Do not alter value)
    NonModal_GoHome_1 = 30,                 // G30 (Do not alter value)
    NonModal_SetHome_1 = 40,                // G30.1 (Do not alter value)
    NonModal_AbsoluteOverride = 53,         // G53 (Do not alter value)
    NonModal_SetCoordinateOffset = 92,      // G92 (Do not alter value)
    NonModal_ResetCoordinateOffset = 102,   // G92.1 (Do not alter value)
    NonModal_ClearCoordinateOffset = 112,   // G92.2 (Do not alter value)
    NonModal_RestoreCoordinateOffset = 122  // G92.3 (Do not alter value)
} non_modal_t;

// Modal Group G1: Motion modes
typedef enum {
    MotionMode_Seek = 0,                    // G0 (Default: Must be zero)
    MotionMode_Linear = 1,                  // G1 (Do not alter value)
    MotionMode_CwArc = 2,                   // G2 (Do not alter value)
    MotionMode_CcwArc = 3,                  // G3 (Do not alter value)
    MotionMode_SpindleSynchronized = 33,    // G33 (Do not alter value)
    MotionMode_DrillChipBreak = 73,         // G73 (Do not alter value)
    MotionMode_Threading = 76,              // G76 (Do not alter value)
    MotionMode_CannedCycle81 = 81,          // G81 (Do not alter value)
    MotionMode_CannedCycle82 = 82,          // G82 (Do not alter value)
    MotionMode_CannedCycle83 = 83,          // G83 (Do not alter value)
    MotionMode_CannedCycle85 = 85,          // G85 (Do not alter value)
    MotionMode_CannedCycle86 = 86,          // G86 (Do not alter value)
    MotionMode_CannedCycle89 = 89,          // G89 (Do not alter value)
    MotionMode_ProbeToward = 140,           // G38.2 (Do not alter value)
    MotionMode_ProbeTowardNoError = 141,    // G38.3 (Do not alter value)
    MotionMode_ProbeAway = 142,             // G38.4 (Do not alter value)
    MotionMode_ProbeAwayNoError = 143,      // G38.5 (Do not alter value)
    MotionMode_None = 80                    // G80 (Do not alter value)
} motion_mode_t;

// Modal Group G2: Plane select
typedef enum {
    PlaneSelect_XY = 0, // G17 (Default: Must be zero)
    PlaneSelect_ZX = 1, // G18 (Do not alter value)
    PlaneSelect_YZ = 2  // G19 (Do not alter value)
} plane_select_t;

// Modal Group G4: Arc IJK distance mode
//#define DISTANCE_ARC_MODE_INCREMENTAL 0 // G91.1 (Default: Must be zero)

// Modal Group M4: Program flow
typedef enum {
    ProgramFlow_Running = 0,        // (Default: Must be zero)
    ProgramFlow_Paused  = 3,        // M0
    ProgramFlow_OptionalStop = 1,   // M1
    ProgramFlow_CompletedM2 = 2,    // M2 (Do not alter value)
    ProgramFlow_CompletedM30 = 30   // M30 (Do not alter value)
} program_flow_t;

// Modal Group G5: Feed rate mode
typedef enum {
    FeedMode_UnitsPerMin = 0,   // G94 (Default: Must be zero)
    FeedMode_InverseTime = 1,   // G93 (Do not alter value)
    FeedMode_UnitsPerRev = 2    // G95 (Do not alter value)
} feed_mode_t;

// Modal Group G10: Canned cycle return mode
typedef enum {
    CCRetractMode_Previous = 0,  // G98 (Default: Must be zero)
    CCRetractMode_RPos = 1       // G99 (Do not alter value)
} cc_retract_mode_t;

// Modal Group G7: Cutter radius compensation mode
//#define CUTTER_COMP_DISABLE 0 // G40 (Default: Must be zero)

// Modal Group G13: Control mode
//#define CONTROL_MODE_EXACT_PATH 0 // G61 (Default: Must be zero)

// Modal Group G8: Tool length offset
typedef enum {
    ToolLengthOffset_Cancel = 0,         // G49 (Default: Must be zero)
    ToolLengthOffset_Enable = 1,         // G43
    ToolLengthOffset_EnableDynamic = 2,  // G43.1
    ToolLengthOffset_ApplyAdditional = 3 // G43.2
} tool_offset_mode_t;

// Modal Group M9: Override control
typedef enum {
    Override_Parking = 56,          // M56
    Override_FeedHold = 53,         // M53
    Override_FeedSpeed = 49,        // M49
    Override_FeedRate = 50,         // M50
    Override_SpindleSpeed = 51      // M51
} override_mode_t;

// Modal Group G12: Active work coordinate system
// N/A: Stores coordinate system value (54-59) to change to.

// Modal Group G14: Spindle Speed Mode
typedef enum {
    SpindleSpeedMode_RPM = 0,  // G96 (Default: Must be zero)
    SpindleSpeedMode_CSS = 1   // G97 (Do not alter value)
} spindle_rpm_mode_t;

// Modal Group M10: User defined M commands
// NOTE: Not used by core, may be used by driver code
typedef enum {
    UserMCode_Ignore = 0,
    Trinamic_DebugReport = 122,
    Trinamic_StepperCurrent = 906,
    Trinamic_ModeToggle = 569,
    Trinamic_ReportPrewarnFlags = 911,
    Trinamic_ClearPrewarnFlags = 912,
    Trinamic_HybridThreshold = 913,
    Trinamic_HomingSensivity = 914
} user_mcode_t;

// Define g-code parser position updating flags
typedef enum {
    GCUpdatePos_Target = 0,
    GCUpdatePos_System,
    GCUpdatePos_None
} pos_update_t;

// Define probe cycle exit states and assign proper position updating.
typedef enum {
    GCProbe_Found = GCUpdatePos_System,
    GCProbe_Abort = GCUpdatePos_None,
    GCProbe_FailInit = GCUpdatePos_None,
    GCProbe_FailEnd = GCUpdatePos_Target,
  #ifdef SET_CHECK_MODE_PROBE_TO_START
    GCProbe_CheckMode = GCUpdatePos_None
  #else
    GCProbe_CheckMode = GCUpdatePos_Target
  #endif
} gc_probe_t;

// Define gcode parser flags for handling special cases.

typedef union {
    uint16_t value;
    struct {
        uint16_t jog_motion          :1,
                 canned_cycle_change :1,
                 arc_is_clockwise    :1,
                 probe_is_away       :1,
                 probe_is_no_error   :1,
                 spindle_force_sync  :1,
                 laser_disable       :1,
                 laser_is_motion     :1,
                 set_coolant         :1,
                 reserved            :7;
    };
} gc_parser_flags_t;

typedef union {
    uint8_t value;
    struct {
        uint8_t feed_rate_disable   :1,
                feed_hold_disable   :1,
                spindle_rpm_disable :1,
                parking_disable     :1,
                reserved            :3,
                sync                :1;
    };
} gc_override_flags_t;

typedef union {
    float values[N_AXIS];
    struct {
        float x;
        float y;
        float z;
#ifdef A_AXIS
        float a;
#endif
#ifdef B_AXIS
        float b;
#endif
#ifdef C_AXIS
        float c;
#endif
    };
} coord_data_t;

typedef struct {
    float xyz[N_AXIS];
    uint8_t idx;
} coord_system_t;

typedef struct {
    uint8_t axis_0;
    uint8_t axis_1;
    uint8_t axis_linear;
} plane_t;

// NOTE: When this struct is zeroed, the above defines set the defaults for the system.
typedef struct {
    motion_mode_t motion;                // {G0,G1,G2,G3,G38.2,G80}
    feed_mode_t feed_mode;               // {G93,G94}
    bool units_imperial;                 // {G20,G21}
    bool distance_incremental;           // {G90,G91}
    bool diameter_mode;                  // {G7,G8} Lathe diameter mode
    // uint8_t distance_arc;             // {G91.1} NOTE: Don't track. Only default supported.
    plane_select_t plane_select;         // {G17,G18,G19}
    // uint8_t cutter_comp;              // {G40} NOTE: Don't track. Only default supported.
    tool_offset_mode_t tool_offset_mode; // {G43,G43.1,G49}
    coord_system_t coord_system;         // {G54,G55,G56,G57,G58,G59,G59.1,G59.2,G59.3}
    // uint8_t control;                  // {G61} NOTE: Don't track. Only default supported.
    program_flow_t program_flow;         // {M0,M1,M2,M30}
    coolant_state_t coolant;             // {M7,M8,M9}
    spindle_state_t spindle;             // {M3,M4,M5}
    gc_override_flags_t override_ctrl;   // {M48,M49,M50,M51,M53,M56}
    spindle_rpm_mode_t spindle_rpm_mode; // {G96,G97}
    cc_retract_mode_t retract_mode;      // {G98,G99}
    bool scaling_active;                 // {G50,G51}
    bool canned_cycle_active;
} gc_modal_t;

typedef struct {
    float d;                   // Max spindle RPM in Constant Surface Speed Mode (G96)
    float e;                   // Thread taper length (G76)
    float f;                   // Feed
    float ijk[3];              // I,J,K Axis arc offsets
    float k;                   // G33 distance per revolution
    float p;                   // G10 or dwell parameters
    float q;                   // User defined M-code parameter (G82 peck drilling, not supported)
    float r;                   // Arc radius or retract position
    float s;                   // Spindle speed
    float xyz[N_AXIS];         // X,Y,Z Translational axes
    coord_system_t coord_data; // Coordinate data
    int32_t n;                 // Line number
    uint8_t h;                 // Tool number
    uint8_t t;                 // Tool selection
    uint8_t l;                 // G10 or canned cycles parameters
} gc_values_t;

typedef struct {
    float xyz[3];
    float delta;
    float dwell;
    float prev_position;
    float retract_position; // Canned cycle retract position
    bool rapid_retract;
    bool spindle_off;
    cc_retract_mode_t retract_mode;
    bool change;
} gc_canned_t;

typedef enum {
    Taper_None = 0,
    Taper_Entry,
    Taper_Exit,
    Taper_Both
} gc_taper_type;

typedef struct {
    float pitch;
    float z_final;
    float peak;
    float initial_depth;
    float depth;
    float depth_degression;
    float main_taper_height;
    float end_taper_length;
    float infeed_angle;
    float cut_direction;
    uint_fast16_t spring_passes;
    gc_taper_type end_taper_type;
} gc_thread_data;

typedef struct {
    float offset[N_AXIS];
    float radius;
    uint8_t tool;
} tool_data_t;

// Data used for Constant Surface Speed Mode calculations
typedef struct {
    float surface_speed;    // Surface speed in millimeters/min
    float target_rpm;       // Target RPM at end of movement
    float max_rpm;          // Maximum spindle RPM
    float tool_offset;      // Tool offset
    uint_fast8_t axis;      // Linear (tool) axis
    bool active;
} css_data_t;

typedef struct {
    float rpm;      // RPM
    css_data_t css; // Data used for Constant Surface Speed Mode calculations
} spindle_t;

typedef struct {
    gc_modal_t modal;
    gc_canned_t canned;
    spindle_t spindle;                  // RPM
    float feed_rate;                    // Millimeters/min
    float distance_per_rev;             // Millimeters/rev
    float position[N_AXIS];             // Where the interpreter considers the tool to be at this point in the code
    int32_t line_number;                // Last line number sent
    uint8_t tool_pending;               // Tool to be selected on next M6
    bool file_run;                      // Tracks % command
    bool is_laser_ppi_mode;
    bool is_rpm_rate_adjusted;
    bool tool_change;
    status_code_t last_error;           // last return value from parser
    //
    float g92_coord_offset[N_AXIS];     // Retains the G92 coordinate offset (work coordinates) relative to
                                        // machine zero in mm. Non-persistent. Cleared upon reset and boot.
    float tool_length_offset[N_AXIS];   // Tracks tool length offset value when enabled
    tool_data_t *tool;                  // Tracks tool number and offset
} parser_state_t;

typedef struct {
    float xyz[N_AXIS]; // Center point
    float ijk[N_AXIS]; // Scaling factors
} scale_factor_t;

extern parser_state_t gc_state;
#ifdef N_TOOLS
extern tool_data_t tool_table[N_TOOLS + 1];
#else
extern tool_data_t tool_table;
#endif

typedef struct {
    non_modal_t non_modal_command;
    override_mode_t override_command; // TODO: add to non_modal above?
    user_mcode_t user_mcode;
    bool user_mcode_sync;
    gc_modal_t modal;
    gc_values_t values;
} parser_block_t;

// Initialize the parser
void gc_init(bool cold_start);

// Execute one block of rs275/ngc/g-code
status_code_t gc_execute_block(char *block, char *message);

// Sets g-code parser position in mm. Input in steps. Called by the system abort and hard
// limit pull-off routines.
#define gc_sync_position() system_convert_array_steps_to_mpos (gc_state.position, sys_position)

void gc_set_laser_ppimode (bool on);

// Gets axes scaling state.
axes_signals_t gc_get_g51_state (void);
float *gc_get_scaling (void);

// Get current axis offset.
float gc_get_offset (uint_fast8_t idx);

#endif
